// amadeus/lib/amadeus/client/errors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnknownError = exports.ServerError = exports.ResponseError = exports.ParserError = exports.NotFoundError = exports.NetworkError = exports.ClientError = exports.AuthenticationError = void 0;
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
/**
 * The error that is passed to the Promise when the API call fails.
 *
 * @param {Response} response the {@link Response} object containing the raw
 *  http response and the {@link Request} instance that made the API call.
 * @property {Response} response the {@link Response} object containing the raw
 *  http response and the {@link Request} instance that made the API call.
 * @property {string} code a unique code for this type of error. Options include
 *  `NetworkError`, `ParserError`, `ResponseError`, `ServerError`,
 *  `AuthenticationError`, `NotFoundError` and `UnknownError`
 *  from the  {@link Response}'s parsed data
 */
var ResponseError = /*#__PURE__*/function () {
  function ResponseError(response) {
    _classCallCheck(this, ResponseError);
    this.response = response;
    this.determineDescription();
  }

  // PRIVATE
  _createClass(ResponseError, [{
    key: "determineDescription",
    value: function determineDescription() {
      if (!this.response || !this.response.parsed) {
        this.description = null;
        return;
      }
      var result = this.response.result;
      if (result && result.errors) {
        this.description = result.errors;
      } else if (result) {
        this.description = result;
      }
      return;
    }
  }]);
  return ResponseError;
}(); // Protected
exports.ResponseError = ResponseError;
var NetworkError = /*#__PURE__*/function (_ResponseError) {
  _inherits(NetworkError, _ResponseError);
  var _super = _createSuper(NetworkError);
  function NetworkError() {
    var _this;
    _classCallCheck(this, NetworkError);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    _this.code = 'NetworkError';
    return _this;
  }
  return _createClass(NetworkError);
}(ResponseError);
exports.NetworkError = NetworkError;
var ParserError = /*#__PURE__*/function (_ResponseError2) {
  _inherits(ParserError, _ResponseError2);
  var _super2 = _createSuper(ParserError);
  function ParserError() {
    var _this2;
    _classCallCheck(this, ParserError);
    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }
    _this2 = _super2.call.apply(_super2, [this].concat(args));
    _this2.code = 'ParserError';
    return _this2;
  }
  return _createClass(ParserError);
}(ResponseError);
exports.ParserError = ParserError;
var ServerError = /*#__PURE__*/function (_ResponseError3) {
  _inherits(ServerError, _ResponseError3);
  var _super3 = _createSuper(ServerError);
  function ServerError() {
    var _this3;
    _classCallCheck(this, ServerError);
    for (var _len3 = arguments.length, args = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
      args[_key3] = arguments[_key3];
    }
    _this3 = _super3.call.apply(_super3, [this].concat(args));
    _this3.code = 'ServerError';
    return _this3;
  }
  return _createClass(ServerError);
}(ResponseError);
exports.ServerError = ServerError;
var ClientError = /*#__PURE__*/function (_ResponseError4) {
  _inherits(ClientError, _ResponseError4);
  var _super4 = _createSuper(ClientError);
  function ClientError() {
    var _this4;
    _classCallCheck(this, ClientError);
    for (var _len4 = arguments.length, args = new Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
      args[_key4] = arguments[_key4];
    }
    _this4 = _super4.call.apply(_super4, [this].concat(args));
    _this4.code = 'ClientError';
    return _this4;
  }
  return _createClass(ClientError);
}(ResponseError);
exports.ClientError = ClientError;
var AuthenticationError = /*#__PURE__*/function (_ResponseError5) {
  _inherits(AuthenticationError, _ResponseError5);
  var _super5 = _createSuper(AuthenticationError);
  function AuthenticationError() {
    var _this5;
    _classCallCheck(this, AuthenticationError);
    for (var _len5 = arguments.length, args = new Array(_len5), _key5 = 0; _key5 < _len5; _key5++) {
      args[_key5] = arguments[_key5];
    }
    _this5 = _super5.call.apply(_super5, [this].concat(args));
    _this5.code = 'AuthenticationError';
    return _this5;
  }
  return _createClass(AuthenticationError);
}(ResponseError);
exports.AuthenticationError = AuthenticationError;
var NotFoundError = /*#__PURE__*/function (_ResponseError6) {
  _inherits(NotFoundError, _ResponseError6);
  var _super6 = _createSuper(NotFoundError);
  function NotFoundError() {
    var _this6;
    _classCallCheck(this, NotFoundError);
    for (var _len6 = arguments.length, args = new Array(_len6), _key6 = 0; _key6 < _len6; _key6++) {
      args[_key6] = arguments[_key6];
    }
    _this6 = _super6.call.apply(_super6, [this].concat(args));
    _this6.code = 'NotFoundError';
    return _this6;
  }
  return _createClass(NotFoundError);
}(ResponseError);
exports.NotFoundError = NotFoundError;
var UnknownError = /*#__PURE__*/function (_ResponseError7) {
  _inherits(UnknownError, _ResponseError7);
  var _super7 = _createSuper(UnknownError);
  function UnknownError() {
    var _this7;
    _classCallCheck(this, UnknownError);
    for (var _len7 = arguments.length, args = new Array(_len7), _key7 = 0; _key7 < _len7; _key7++) {
      args[_key7] = arguments[_key7];
    }
    _this7 = _super7.call.apply(_super7, [this].concat(args));
    _this7.code = 'UnknownError';
    return _this7;
  }
  return _createClass(UnknownError);
}(ResponseError);
exports.UnknownError = UnknownError;
